### PR TITLE
Use volume textures for LUT filter

### PR DESCRIPTION
**Pull request by jpark37**
### Description
Adds support for volume textures to the renderer, and uses them for the LUT filter

### Motivation and Context
Silly to perform manual calculations to treat 2D texture as 3D texture.

### How Has This Been Tested?
Tested D3D and OpenGL (Windows, Ubuntu WSL). LUT output matches master branch, give or take epsilon.

Not much perf gain, but GPA shows bottleneck shift from Shader Execution to Sampler as expected.

### Types of changes
- New feature (non-breaking change which adds functionality)
- Performance enhancement (non-breaking change which improves efficiency)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

### CODE REVIEW

File: plugins/obs-filters/color-grade-filter.c [hunk 1]
```diff
@@ -28,6 +32,54 @@ static const char *color_grade_filter_get_name(void *unused)
...
+				       const uint8_t *data)
+{
+	if (image_width % LUT_WIDTH != 0)
+		return NULL;
+
+	if (image_height % LUT_WIDTH != 0)
+		return NULL;
+
+	const uint32_t pixel_count = LUT_WIDTH * LUT_WIDTH * LUT_WIDTH;
+	if ((image_width * image_height) != pixel_count)
```
Comment (Xaymar) [inline plugins/obs-filters/color-grade-filter.c:47]: This breaks lookup tables that aren't exactly 64x64x64, such as a LUT that is 256x256x256 and would cover the entire RGB color spectrum. (Which currently aren't supported anyway)

Comment (Xaymar) [reply plugins/obs-filters/color-grade-filter.c:47]: LUT images that could work if supported:

* 1x1x1: Basically useless.
* 4x4x4: 2D Size 8x8. Useful for quick and dirty grayscale, interpolation makes up the majority of quality. Very very fast for LUT.
* 16x16x16: 2D Size 64x64. Higher quality Grayscale and isolation at the cost of more sampling.
* 64x64x64: The current size of 512x512, covers 1.56% of the representable spectrum, making up the rest of interpolation.
* 256x256x256: 4096x4096, **covers the entire 8-bit color range**.
* 1024x1024x1024: 32768x32768, **covers the entire 10-bit color range**, will definitely require volume textures as the U/V limits for 2D textures cap out at 16384 - same for OpenGL it seems. 3D textures can go as big as 2048x2048x2048, which would cover 11-bits, but nothing uses 11-bits.

**Comment by Xaymar** [comment]
This PR seems to contain many unrelated changes that should be a separate PR (the entire changes to gs, libobs-d3d11 and libobs-opengl).

I can see two gains from using volume textures over standard textures:

- Effect code is cleaner.
- 10-bit Color Range is supported (thanks to the UVW limit of >1024), which is not possible with 2D textures (UV limit of <=16384).

More effective gains in performance could be achieved by adding support for additional LUT sizes, see my other comment. Grayscale and color isolation/muting don't need the full 64x64x64, but posterization for example would need at least 64x64x64.

File: plugins/obs-filters/color-grade-filter.c (hunk 1 above)
Comment (jpark37) [reply plugins/obs-filters/color-grade-filter.c]: LUTs that aren't 64x64x64 were not previously supported. The shader was hard-coded for them.

**Comment by jpark37** [comment]
I was going to wait until another PR to make the LUT size actually flexible, and add support for Adobe CUBE files in the process. This just starts the process of refactoring the code in that direction.

File: plugins/obs-filters/color-grade-filter.c (hunk 1 above)
Comment (Xaymar) [reply plugins/obs-filters/color-grade-filter.c:47]: Yeah, I realized in my 2nd comment.

**Comment by jpark37** [comment]
Rebased locally onto #2264 to test Mac, and this looks fine. Marking as ready.

Also generated grayscale.png because black_and_white.png appears to be something else, and made me think I might have added a regression.

image

**Pull request merged**